group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1307,17 +1305,25 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
             }
         }
 
-        // The subscription metadata is updated in two cases:
-        // 1) The member has updated its subscriptions;
-        // 2) The refresh deadline has been reached.
+        int groupEpoch = group.groupEpoch();
+        Map<String, TopicMetadata> subscriptionMetadata = group.subscriptionMetadata();
         Map<String, Integer> subscribedTopicNamesMap = group.subscribedTopicNames();
+        SubscriptionType subscriptionType = group.subscriptionType();
+
         if (bumpGroupEpoch || group.hasMetadataExpired(currentTimeMs)) {
+            // The subscription metadata is updated in two cases:
+            // 1) The member has updated its subscriptions;
+            // 2) The refresh deadline has been reached.
             subscribedTopicNamesMap = group.computeSubscribedTopicNames(member, updatedMember);
             subscriptionMetadata = group.computeSubscriptionMetadata(
                 subscribedTopicNamesMap,
                 metadataImage.topics(),
                 metadataImage.cluster()
             );
+            subscriptionType = ConsumerGroup.subscriptionType(
+                subscribedTopicNamesMap,
+                group.numMembers()
```

Comment:
Once https://github.com/apache/kafka/pull/15847 is merged, I think that we can easily compute the numMembers as follow:

```
int numMembers = group.numMembers();
if (!group.hasMember(updatedMember.memberId() && !staticMemberReplaced) {
   numMembers++;
}
```

If the member did not exist and we are not replacing a static member, we increment by one.